### PR TITLE
Feat(content): improve deterrence calculation by taking special damages into account

### PR DESCRIPTION
**Pull request by Hurleveur**
**Feature:** This PR implements the feature request detailed and discussed in issue (sort of) #7894 notably

## Feature Details
Justification for each kind of special effect added:
Take corrosion and discharge into account because they affect hull and shields directly.
Take disruption as well because it is very potent to skip shields.

Also consider energy damages and the like given they can render your ship a sitting duck and have an effect before that even happens, by reducing your fire output, ability to escape etc
Finally, consider heat since it has the potential to render you a sitting duck, although it does not matter as much.

## UI Screenshots
n/a

## Usage Examples
I took a look at some weapons like plasma, inferno, ion cannon etc and they have more "normal" deterrence now instead of being nearly useless on that front.
I'm willing to change the exact coefficient but we can always try it and see how it goes, surely it can't be more broken than it currently is where it completely ignores those things.

## Testing Done
none

### Automated Tests Added
n/a

## Performance Impact
n/a since I changed it to only calculate when it has to recently

### CODE REVIEW

File: source/Ship.cpp [hunk 1]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
 			const Outfit *weapon = hardpoint.GetOutfit();
 			if(weapon->Ammo() && weapon->AmmoUsage() && !OutfitCount(weapon->Ammo()))
 				continue;
-			double strength = weapon->ShieldDamage() + weapon->HullDamage()
-				+ (weapon->RelativeShieldDamage() * attributes.Get("shields"))
-				+ (weapon->RelativeHullDamage() * attributes.Get("hull"));
+			// Ignore disabled and asteroid damages because they don't apply here.
+			// Ignore fuel and slowing damage because they don't affect combat ability.
+			// Over time effects are per frame but dont affect instantaniously so only multiply by 30.
```
Comment (Quantumshark) [inline source/Ship.cpp:4406]:
```suggestion
			// Over time effects are per frame but dont affect instantaneously so only multiply by 30.
```

File: source/Ship.cpp [hunk 2]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+			// Burning and ion do not affect ships as much, but ion does so more than heat.
+			double strength = (weapon->ShieldDamage() + weapon->RelativeShieldDamage() * attributes.Get("shields"))
+				+ weapon->DischargeDamage() * 30.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+				+ weapon->CorrosionDamage() * 30.
+				+ weapon->DisruptionDamage() * 30.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy") / 2.)
```
Comment (Quantumshark) [inline source/Ship.cpp:4413]:
```suggestion
				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity") / 2.)
```

File: source/Ship.cpp [hunk 3]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+				+ weapon->IonDamage() * 30. + weapon->DisruptionDamage() * 30.) / 4.
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * attributes.Get("heat"))
```
[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/Ship.cpp:4415]:
```suggestion
				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * attributes.Get("heat capacity"))
```

File: source/Ship.cpp [hunk 4]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+			// Burning and ion do not affect ships as much, but ion does so more than heat.
+			double strength = (weapon->ShieldDamage() + weapon->RelativeShieldDamage() * attributes.Get("shields"))
+				+ weapon->DischargeDamage() * 30.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+				+ weapon->CorrosionDamage() * 30.
+				+ weapon->DisruptionDamage() * 30.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity") / 2.)
+				+ weapon->IonDamage() * 30. + weapon->DisruptionDamage() * 30.) / 4.
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * attributes.Get("heat capacity"))
+				+ weapon->BurnDamage() * 30.) / 6.;
```
Comment (UnorderedSigh) [inline source/Ship.cpp:4416]:
```suggestion
				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity") / 2.)
					+ weapon->IonDamage() * 30. + weapon->DisruptionDamage() * 30.) / 4.
				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * attributes.Get("heat capacity"))
					+ weapon->BurnDamage() * 30.) / 6.;
```

Additional indentation in the nested parentheses blocks will make this easier to follow.

**Review by UnorderedSigh — APPROVED** [review]
This better represents the effectiveness of Korath weapons. Their burn and corrosion damage can be quite deadly, which wasn't considered at all before.

File: source/Ship.cpp [hunk 5]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+			// Over time effects are per frame but dont affect instantaneously so only multiply by 30.
+			// Burning and ion do not affect ships as much, but ion does so more than heat.
+			double strength = (weapon->ShieldDamage() + weapon->RelativeShieldDamage() * attributes.Get("shields"))
+				+ weapon->DischargeDamage() * 30.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+				+ weapon->CorrosionDamage() * 30.
+				+ weapon->DisruptionDamage() * 30.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity") / 2.)
+					+ weapon->IonDamage() * 30. + weapon->DisruptionDamage() * 30.) / 4.
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * attributes.Get("heat capacity"))
```
[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/Ship.cpp:4415]: I cant look rn but doesnt the new frugal test make the same error?

[the remark this answers is not in the document]
Comment (warp-core) [reply source/Ship.cpp:4415]: Not that I can tell.
The `heat capacity` attribute is only ever retrieved in one place: Ship::MaximumHeat.

[the remark this answers is not in the document]
Comment (Hurleveur) [reply source/Ship.cpp:4415]:
```suggestion
				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * MaximumHeat())
```

File: source/Ship.cpp [hunk 6]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
 			const Outfit *weapon = hardpoint.GetOutfit();
 			if(weapon->Ammo() && weapon->AmmoUsage() && !OutfitCount(weapon->Ammo()))
 				continue;
-			double strength = weapon->ShieldDamage() + weapon->HullDamage()
-				+ (weapon->RelativeShieldDamage() * attributes.Get("shields"))
-				+ (weapon->RelativeHullDamage() * attributes.Get("hull"));
+			// Ignore disabled and asteroid damages because they don't apply here.
+			// Ignore fuel and slowing damage because they don't affect combat ability.
```
Comment (warp-core) [inline source/Ship.cpp:4405]: That's just not true?
If the ship in question relies on Flamethrowers, or other fuel consuming weapons, fuel damage will absolutely affect its combat ability.
For slowing damage, having a significantly reduced turn rate will definitely affect the combat ability of a ship that relies on fixed guns, especially if the projectiles do not have good homing or manoeuvrability.
Also, no mention of leak?

File: source/Ship.cpp [hunk 7]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+			// Over time effects are per frame but dont affect instantaneously so only multiply by 30.
```
Comment (warp-core) [inline source/Ship.cpp:4406]: What's the basis for this?
Applying 10 discharge to a ship will result in the loss of 1000 shields over the course of the discharge's lifetime.
Why does the fact that it is spread out sufficient to make it only as deterring as 300 shield damage?
Why is disruption being multiplied by the same value? It's quite different a mechanic.

File: source/Ship.cpp [hunk 8]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+			// Burning and ion do not affect ships as much, but ion does so more than heat.
```
Comment (warp-core) [inline source/Ship.cpp:4407]: I'm not sure what this sentence is saying.

File: source/Ship.cpp [hunk 9]
```diff
@@ -4401,9 +4401,19 @@ double Ship::CalculateDeterrence() const
...
+			double strength = (weapon->ShieldDamage() + weapon->RelativeShieldDamage() * attributes.Get("shields"))
+				+ weapon->DischargeDamage() * 30.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+				+ weapon->CorrosionDamage() * 30.
+				+ weapon->DisruptionDamage() * 30.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity") / 2.)
+					+ weapon->IonDamage() * 30. + weapon->DisruptionDamage() * 30.) / 4.
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * MaximumHeat())
+					+ weapon->BurnDamage() * 30.) / 6.;
```
Comment (warp-core) [inline source/Ship.cpp:4416]: This is not what I'd describe as "readable."
Also, why is there disruption damage in the energy/ion damage line?

File: source/Ship.cpp (hunk 6 above)
Comment (Hurleveur) [reply source/Ship.cpp:4405]: ~~as if we're going to take all of that into account~~
we could at some point but this is already a big improvement, although let me rephrase that
```suggestion
			// Ignore fuel, leak and slowing damage for now, as their affect on combat abilities varies.
```
I mean we consider weapons that dont have gas to fire for scaring pirates anyway - how would I measure all of this stuff tho, feels too complex for this PR

File: source/Ship.cpp (hunk 8 above)
Comment (Hurleveur) [reply source/Ship.cpp:4407]: ion's worse than heat because the relative effect of 1 ion on a ship is worse than 1 heat for pirates in general?

File: source/Ship.cpp (hunk 9 above)
Comment (Hurleveur) [reply source/Ship.cpp:4416]: you're right it could be made more readable
no particular reason so I'll move it ig

File: source/Ship.cpp (hunk 7 above)
Comment (Hurleveur) [reply source/Ship.cpp:4406]: ~~because I dont truly understand all of these and thought it'd be a basis to improve upon since the base formula sucks so it cant be much worse~~
1000? damn that's a lot, but how long is said lifetime?

File: source/Ship.cpp [hunk 10]
```diff
@@ -4401,9 +4401,20 @@ double Ship::CalculateDeterrence() const
...
+				+ weapon->SlowingDamage() * 60.
+				+ (weapon->FuelDamage() + weapon->LeakDamage() * 60)
+				+ (weapon->ShieldDamage() + weapon->RelativeShieldDamage() * attributes.Get("shields"))
+					+ weapon->DischargeDamage() * 60.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+					+ weapon->CorrosionDamage() * 60.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity"))
+					+ weapon->IonDamage() * 60. + weapon->ScramblingDamage() * 60) / 2.
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * MaximumHeat())
+					+ weapon->BurnDamage() * 60.) / 10.;
```
Comment (Amazinite) [inline source/Ship.cpp:4417]: We're effectively giving every damage type a similar rating with these multipliers, which I don't think makes much sense. For example, fuel damage can be a deterrent, but how often will fuel damage actually influence the outcome of a battle? Extremely few ships use fuel damage weapons.

I think that the primary means of deterrence should be shield and hull damage. If we include special damage types, then the influence that those special damage types have on the deterrence strength should be lower.

**Review by Amazinite — CHANGES_REQUESTED** [review]

File: source/Ship.cpp (hunk 10 above)
Comment (Quantumshark) [reply source/Ship.cpp:4417]: I would consider fuel damage to be a deterrent, not because it's likely to be the difference between victory and defeat, but because it's likely to be the difference between escape and destruction. The ability to eliminate a ship's ability to escape is a serious threat, even if it's unlikely to directly influence a hypothetical duel.

Comment (Hurleveur) [reply source/Ship.cpp:4417]: well the influence of those damage types is lower already for all weapons pretty much - heat gets divided by 10 because its easy to get a big amount of that, but for other types 100 slowing is a lot? I did this with different weapons in mind, trying to make them have an appropriate amount of deterrance but we can tweak it for sure

File: source/Ship.cpp [hunk 11]
```diff
@@ -4574,9 +4574,22 @@ double Ship::CalculateDeterrence() const
...
+					+ weapon->DischargeDamage() * 60.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+					+ weapon->CorrosionDamage() * 60.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity"))
+					+ weapon->IonDamage() * 60. + weapon->ScramblingDamage() * 60) * .25
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * MaximumHeat())
+					+ weapon->BurnDamage() * 60.) * .1
+				+ weapon->SlowingDamage() * 10.
+				+ (weapon->FuelDamage() + weapon->LeakDamage() * 60.) * 2.;
+
```
Comment (Hurleveur) [inline source/Ship.cpp:4592]:
```suggestion

```

File: source/Ship.cpp [hunk 12]
```diff
@@ -4574,9 +4574,21 @@ double Ship::CalculateDeterrence() const
...
+				+ (weapon->ShieldDamage() + weapon->RelativeShieldDamage() * attributes.Get("shields"))
+					+ weapon->DischargeDamage() * 60.
+				+ (weapon->HullDamage() + weapon->RelativeHullDamage() * attributes.Get("hull"))
+					+ weapon->CorrosionDamage() * 60.
+				+ ((weapon->EnergyDamage() + weapon->RelativeEnergyDamage() * attributes.Get("energy capacity"))
+					+ weapon->IonDamage() * 60. + weapon->ScramblingDamage() * 60) * .25
+				+ ((weapon->HeatDamage() + weapon->RelativeHeatDamage() * MaximumHeat())
+					+ weapon->BurnDamage() * 60.) * .1
+				+ weapon->SlowingDamage() * 10.
+				+ (weapon->FuelDamage() + weapon->LeakDamage() * 60.) * 2.;
```
Comment (Amazinite) [inline source/Ship.cpp:4591]: Breaking up this variable a bit to make it easier to read.
Multiplying DoT effects by 100 instead of by 60 because that's the actual conversion between DoT applied to damage dealt. If you get hit with 1 corrosion damage then you will eventually feel 
Applying a blanket 0.2x multiplier to special damage types instead of some being .25, .2, or .1.
You missed relative fuel damage.
```suggestion
			// 1 DoT damage of type X = 100 damage of type X over an extended period of time
			// (~95 damage after 5 seconds, ~99 damage after 8 seconds). Therefore, multiply
			// DoT damage types by 100. Disruption, scrambling, and slowing don't have an
			// analogous instantaneous damage type, but still just multiply them by 100 to
			// stay consistent.

			// Compare the relative damage types to the strength of the firing ship, since we
			// have nothing else to reasonably compare against.

			// Shield and hull damage are the primary damage types that dictate combat, so
			// consider the full damage dealt by these types for the strength of a weapon.
			double shieldFactor = weapon->ShieldDamage()
					+ weapon->RelativeShieldDamage() * attributes.Get("shields")
					+ weapon->DischargeDamage() * 100.
			double hullFactor = weapon->HullDamage()
					+ weapon->RelativeHullDamage() * attributes.Get("hull")
					+ weapon->CorrosionDamage() * 100.;

			// Other damage types don't outright destroy ships, so they aren't considered
			// as heavily in the strength of a weapon.
			double energyFactor = weapon->EnergyDamage()
					+ weapon->RelativeEnergyDamage() * attributes.Get("energy capacity")
					+ weapon->IonDamage() * 100.;
			double heatFactor = weapon->HeatDamage()
					+ weapon->RelativeHeatDamage() * MaximumHeat()
					+ weapon->BurnDamage() * 100.;
			double fuelFactor = weapon->FuelDamage()
					+ weapon->RelativeFuelDamage() * attributes.Get("fuel capacity")
					+ weapon->LeakDamage() * 100.;
			double scramblingFactor = weapon->ScramblingDamage() * 100.;
			double slowingFactor = weapon->SlowingDamage() * 100.;
			double disruptionFactor = weapon->DisruptionDamage() * 100.;

			// Disabled and asteroid damage are ignored because they don't matter in combat.

			double strength = shieldFactor + hullFactor + 0.2 * (energyFactor + heatFactor + fuelFactor
					+ scramblingFactor + slowingFactor + disruptionFactor);
```

**Review by Amazinite — APPROVED** [review]
